plugins/database/redshift/redshift.go

```diff
@@ -99,124 +112,54 @@ func (r *RedShift) getConnection(ctx context.Context) (*sql.DB, error) {
+// NewUser creates a new user in the database. There is no default statement for
+// creating users, so one must be specified in the plugin config.
+// Generated usernames are of the form v-{display-name}-{role-name}-{UUID}-{timestamp}
+func (r *RedShift) NewUser(ctx context.Context, req dbplugin.NewUserRequest) (dbplugin.NewUserResponse, error) {
+	if len(req.Statements.Commands) == 0 {
+		return dbplugin.NewUserResponse{}, dbutil.ErrEmptyCreationStatement
 	}
 
 	// Grab the lock
 	r.Lock()
 	defer r.Unlock()
 
-	// Get the connection
-	db, err := r.getConnection(ctx)
-	if err != nil {
-		return "", "", err
+	usernameOpts := []credsutil.UsernameOpt{
+		credsutil.DisplayName(req.UsernameConfig.DisplayName, 8),
+		credsutil.RoleName(req.UsernameConfig.RoleName, 8),
+		credsutil.MaxLength(63),
+		credsutil.Separator("-"),
 	}
-	defer db.Close()
-
-	// Check if the role exists
-	var exists bool
-	err = db.QueryRowContext(ctx, "SELECT exists (SELECT usename FROM pg_user WHERE usename=$1);", username).Scan(&exists)
-	if err != nil && err != sql.ErrNoRows {
-		return "", "", err
+	if r.lowerCaseUsername {
+		usernameOpts = append(usernameOpts, credsutil.ToLower())
 	}
 
-	// Vault requires the database user already exist, and that the credentials
-	// used to execute the rotation statements has sufficient privileges.
-	stmts := statements.Rotation
-
-	// Start a transaction
-	tx, err := db.BeginTx(ctx, nil)
+	username, err := credsutil.GenerateUsername(usernameOpts...)
 	if err != nil {
-		return "", "", err
-	}
-	defer func() {
-		tx.Rollback()
-	}()
-
-	// Execute each query
-	for _, stmt := range stmts {
-		for _, query := range strutil.ParseArbitraryStringSlice(stmt, ";") {
-			query = strings.TrimSpace(query)
-			if len(query) == 0 {
-				continue
-			}
-
-			m := map[string]string{
-				"name":     staticUser.Username,
-				"password": password,
-			}
-			if err := dbtxn.ExecuteTxQuery(ctx, tx, m, query); err != nil {
-				return "", "", err
-			}
-		}
-	}
-
-	// Commit the transaction
-	if err := tx.Commit(); err != nil {
-		return "", "", err
-	}
-	return username, password, nil
-}
-
-func (r *RedShift) CreateUser(ctx context.Context, statements dbplugin.Statements, usernameConfig dbplugin.UsernameConfig, expiration time.Time) (username string, password string, err error) {
-	statements = dbutil.StatementCompatibilityHelper(statements)
-
-	if len(statements.Creation) == 0 {
-		return "", "", dbutil.ErrEmptyCreationStatement
-	}
-
-	// Grab the lock
-	r.Lock()
-	defer r.Unlock()
-
-	username, err = r.GenerateUsername(usernameConfig)
-	if err != nil {
-		return "", "", err
-	}
-
-	password, err = r.GeneratePassword()
-	if err != nil {
-		return "", "", err
-	}
-
-	expirationStr, err := r.GenerateExpiration(expiration)
-	if err != nil {
-		return "", "", err
+		return dbplugin.NewUserResponse{}, err
 	}
+	password := req.Password
+	expirationStr := req.Expiration.UTC().Format("2006-01-02 15:04:05")
```

Comment:
In my experience, postgres has been smart enough to handle different timezones correctly by default, so I didn't think this would be a behaviour change, but it's definitely an unnecessary risk to hope that that holds across all deployments of redshift. I've deleted all occurrences.

Reply:
The problem is 1) this is changing the behavior and 2) the output format doesn't include the timezone

Reply:
Oops I misunderstood your comment before. I think it's fixed properly now, thanks for spotting.